core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingDataEditorPanelService.js

```diff
@@ -0,0 +1,165 @@
+                  stateCustomizationArgsService.savedMemento,
+                  false, 'testInteractionInput'));
+              $scope.trainingData = [];
+
+              var _rebuildTrainingData = function() {
+                $scope.trainingData = [];
+                TrainingDataService.getTrainingDataOfAnswerGroup(
+                  answerGroupIndex).forEach(function(answer) {
+                  answerTemplate = (
+                    ExplorationHtmlFormatterService.getAnswerHtml(
+                      answer, stateInteractionIdService.savedMemento,
+                      stateCustomizationArgsService.savedMemento));
+                  $scope.trainingData.push({
+                    answer: answer,
+                    answerTemplate: answerTemplate
+                  });
+                });
+              };
+
+              $scope.init = function() {
+                _rebuildTrainingData();
+                $scope.newAnswerIsResolved = false;
+                $scope.answerSuccessfullyAdded = false;
+              };
+
+              $scope.removeAnswerFromTrainingData = function(answerIndex) {
+                answer = $scope.trainingData[answerIndex].answer;
+                TrainingDataService.removeAnswerFromAnswerGroupTrainingData(
+                  answer, answerGroupIndex);
+                $scope.trainingData.splice(answerIndex, 1);
+              };
+
+              $scope.exit = function() {
+                $uibModalInstance.close();
+              };
+
+              $scope.submitAnswer = function(newAnswer) {
+                $scope.newAnswerIsResolved = false;
+                $scope.answerSuccessfullyAdded = false;
+
+                var interactionId = stateInteractionIdService.savedMemento;
+
+                var rulesServiceName =
+                  AngularNameService.getNameOfInteractionRulesService(
+                    interactionId);
+
+                // Inject RulesService dynamically.
+                var rulesService = $injector.get(rulesServiceName);
+
+                var newAnswerTemplate = (
+                  ExplorationHtmlFormatterService.getAnswerHtml(
+                    newAnswer, stateInteractionIdService.savedMemento,
+                    stateCustomizationArgsService.savedMemento));
+
+                var classificationResult = (
+                  AnswerClassificationService.getMatchingClassificationResult(
+                    _explorationId, _stateName, _state,
+                    newAnswer, rulesService));
+                var newAnswerFeedbackHtml = 'Nothing';
+                var newAnswerOutcomeDest = classificationResult.outcome.dest;
+                if (classificationResult.outcome.hasNonemptyFeedback()) {
+                  newAnswerFeedbackHtml = (
+                    classificationResult.outcome.feedback.getHtml());
+                }
+                if (newAnswerOutcomeDest === _stateName) {
+                  dest = '<em>(try again)</em>';
```

Comment:
I don't see where `dest` is declared or used...

Also we shouldn't be constructing HTML in the JS code. See if you can do a UI transform on this instead, similar to my comment above for newAnswerFeedbackHtml.

Reply:
Actually, I forgot to replace `dest` with `newAnswerOutcomeDest`. So consider `dest` as `newAnswerOutcomeDest`. Again, this is how it was before so I kept it as it is.

Reply:
Fixed.